### PR TITLE
Refactor dependency between head node and its backend

### PR DESCRIPTION
**Pull request by vepadulano**
This class is not really needed, we can use free functions instead. This removes an implicit inter-dependency between the head node and the backend. Now it's only the head node that stores a reference to the backend needed to execute the graph, similarly to RLoopManager

### CODE REVIEW

File: bindings/experimental/distrdf/python/DistRDF/Proxy.py [hunk 1]
```diff
@@ -54,10 +55,28 @@ def execute_graph(node: Node) -> None:
...
+            # Ingredients needed:
+            # - A representation of the computation graph to be executed
+            # - The logical ranges in which the dataset is split
+            # - A callable to build an RDataFrame instance starting from a
+            #   logical range.
             graph_dict = headnode.generate_graph_dict()  # also takes care of pruning the graph
-            generator = ComputationGraphGenerator(headnode, graph_dict)
-            headnode.backend.execute(generator)
+            ranges = headnode.build_ranges()
+            build_rdf_from_range = headnode.generate_rdf_creator()
```
Comment (etejedor) [inline bindings/experimental/distrdf/python/DistRDF/Proxy.py:68]: You took this out of `execute` and also other things like `handle_returned_values` below, what was the purpose?

Also, what does `build_rdf_from_range` hold?

Comment (vepadulano) [reply bindings/experimental/distrdf/python/DistRDF/Proxy.py:68]: `build_rdf_from_range` is a function that takes in input a logical range and outputs an RDataFrame instance. 
Each distributed task needs to know how to do this, for example starting from a `TreeRange` it needs to create a TChain that reads the correct files (and is restricted to processing only the wanted entries), then the TChain is wrapped by an RDataFrame that will be the starting node of the execution of the computation graph in the task

Comment (vepadulano) [reply bindings/experimental/distrdf/python/DistRDF/Proxy.py:68]: Maybe an alternative for the `BaseBackend.execute` function

```diff
-def execute(self, graph: Dict[int, "Node"], ranges: Iterable, build_rdf_from_range: Callable) -> Iterable:
+def execute(self, headnode: HeadNode) -> Iterable:
```

Taking directly the headnode, no implicit dependency between the classes. Then leaving all the functions of the headnode that were called in the `execute` function could still stay there

**Comment by vepadulano** [comment]
After some more thought, I decided to directly removed any usage of the headnode internal functions from the backend class and also from the Proxy module. The head node already had all the information needed to represent the computation graph locally, split the dataset in logical ranges,  reconstruct and execute the graph in the distributed tasks. Taking inspiration from the RLoopManager class, the head node now also handles the execution, since anyway it has a reference to the backend and everything else needed already embedded. To clarify all the ingredients needed from a head node type, I also made it an abstract class with the methods that subclasses should define

**Review by etejedor — APPROVED** [review]

**Comment by vepadulano** [comment]
Rebased on top of recently merged #10565 , will merge if tests pass

**Pull request merged**